homeassistant/components/habitica.py

```diff
@@ -0,0 +1,134 @@
+"""
+The Habitica API component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/habitica/
+"""
+
+import logging
+from collections import namedtuple
+
+import voluptuous as vol
+from homeassistant.const import \
+    CONF_NAME, CONF_URL, CONF_SENSORS, CONF_PATH
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers import \
+    config_validation as cv, discovery
+
+REQUIREMENTS = ['habitipy==0.2.0']
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "habitica"
+
+CONF_API_USER = "api_user"
+CONF_API_KEY = "api_key"
+
+CONF_EXCLUDE_NAMES = 'exclude_names'
+CONF_SENSORS_LIST = 'sensors_customization'
+ST = SensorType = namedtuple('SensorType', [
+    "name", "icon", "unit", "path"
+])
+
+SENSORS_TYPES = {
+    'name': ST('Name', None, '', ["profile", "name"]),
+    'hp': ST('HP', 'mdi:heart', 'HP', ["stats", "hp"]),
+    'maxHealth': ST('max HP', 'mdi:heart', 'HP', ["stats", "maxHealth"]),
+    'mp': ST('Mana', 'mdi:auto-fix', 'MP', ["stats", "mp"]),
+    'maxMP': ST('max Mana', 'mdi:auto-fix', 'MP', ["stats", "maxMP"]),
+    'exp': ST('EXP', 'mdi:star', 'EXP', ["stats", "exp"]),
+    'toNextLevel': ST(
+        'Next Lvl', 'mdi:star', 'EXP', ["stats", "toNextLevel"]),
+    'lvl': ST(
+        'Lvl', 'mdi:arrow-up-bold-circle-outline', 'Lvl', ["stats", "lvl"]),
+    'gp': ST('Gold', 'mdi:coin', 'Gold', ["stats", "gp"]),
+    'class': ST('Class', 'mdi:sword', '', ["stats", "class"])
+}
+ALL_SENSORS_TYPES = list(SENSORS_TYPES.keys())
+
+INSTANCE_SCHEMA = vol.Schema({
+    vol.Optional(CONF_URL): cv.url,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_API_USER): cv.string,
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Optional(CONF_SENSORS):
+        vol.All(cv.ensure_list, [vol.In(ALL_SENSORS_TYPES)])
+})
+
+
+INSTANCE_LIST_SCHEMA = vol.All(cv.ensure_list, [INSTANCE_SCHEMA])
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: INSTANCE_LIST_SCHEMA
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_API_CALL = 'api_call'
+ATTR_NAME = CONF_NAME
+ATTR_PATH = CONF_PATH
+ATTR_ARGS = "args"
+EVENT_API_CALL_SUCCESS = "_".join((DOMAIN, SERVICE_API_CALL, "success"))
+
+SERVICE_API_CALL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_NAME): str,
+    vol.Required(ATTR_PATH): vol.All(cv.ensure_list, [str]),
+    vol.Optional(ATTR_ARGS): dict
```

Comment:
If possible validate the keys and values of the dict.

Reply:
Requirements will be installed after validation, so it's not possible to depend on module specific requirements in validation functions. Core dependencies are available though.

Reply:
But this is service call schema, so it should be validated after component initialization and waay after module-specific imports. So there might be a chance for it maybe?

Reply:
Right, sorry. Yes, the service schema validation occurs when the service is called, so after component setup and after requirements are installed.

Reply:
Well, I can verify path and name just as it is done right now in service. That args is a bit trickier as we need a way to validate it in `habitipy`, which is still an [open todo in it](https://github.com/ASMfreaK/habitipy/blob/b2fb08eb0a304ee828ac83000106cf4f89ef26df/habitipy/api.py#L404) and [and is not yet implemented](https://github.com/ASMfreaK/habitipy/blob/b2fb08eb0a304ee828ac83000106cf4f89ef26df/habitipy/api.py#L446), sorry.

Reply:
I *think* we can leave it without any validation. With any error it explodes quite loadly in logs. I'll also make it additionally clear in docs that `before you go to write any api calls in HA you should test them in python console using the underlying library`.